### PR TITLE
dataflow: allow batch inserts

### PR DESCRIPTION
**Pull request by benesch**
Adjust SequencedCommand::Insert to allow inserting updates to multiple
local inputs in one shot. In the future, when inserts are persisted to
disk, each insert command will be considered an atomic unit that either
succeeds or fails in its entirety.

### CODE REVIEW

File: src/coord/src/coord.rs [hunk 1]
```diff
@@ -2187,13 +2188,17 @@ impl Coordinator {
                                 if let Some(tables) = &mut self.persisted_tables {
                                     tables.write(id, &updates);
                                 }
-                                self.broadcast(SequencedCommand::Insert { id, updates });
+
+                                updates_map.insert(id, updates);
```
Comment (ruchirK) [inline src/coord/src/coord.rs:2192]: Does it make sense to check that `id` doesn't already exist in `updates_map`. I dunno if it's possible to have that in this context?

File: src/coord/src/coord.rs [hunk 2]
```diff
@@ -409,9 +408,10 @@ impl Coordinator {
                                             // before being fully reloaded.
                                             let updates = std::mem::replace(&mut updates, vec![]);
                                             if !updates.is_empty() {
+                                                let mut updates_map = HashMap::new();
```
Comment (ruchirK) [inline src/coord/src/coord.rs:411]: should I rip all of this code out? @danhhz wdyt -- this seems like unnecessary clutter at this point

**Review by ruchirK — APPROVED** [review]

File: src/dataflow/src/server.rs [hunk 3]
```diff
@@ -93,12 +93,11 @@ pub enum SequencedCommand {
...
     },
-    /// Insert `updates` into the local input named `id`.
+    /// Insert `updates` into local inputs.
     Insert {
         /// Identifier of the local input.
-        id: GlobalId,
-        /// A list of updates to be introduced to the input.
-        updates: Vec<Update>,
+        /// Lists of updates to be introduced to the specified inputs.
+        updates: HashMap<GlobalId, Vec<Update>>,
```
Comment (danhhz) [inline src/dataflow/src/server.rs:100]: nikhil explained offline that this is a HashMap so we can easily construct a session only once per id, but IMO a `Vec<(GlobaId, Vec<Update>)>` is a more natural interface (and makes ruchir's concern above a non-issue). there might be a small performance difference between them (the SequencedCommand::Insert would then either turn the vec into a HashMap or use a HashMap for sessions), but i'd cross that bridge when we come to it. don't feel strongly though

**Review by danhhz — APPROVED** [review]

**Comment by frankmcsherry** [comment]
Can you say a bit about the intended behavioral change of this? Is it roughly the same as introducing `dataflow::Command::MultipleCommands(Vec<dataflow::Command>)`? No moral opposition, but also no opposition to bundling all other dataflow commands up as well.

It seems maybe weird to oblige the `coord <-> dataflow` API to reflect the intended persistence quantum. Could be polite, for sure, but what is the constraint here / why are we changing this API w/o review from coord / dataflow folks?

**Comment by danhhz** [comment]
Hey Frank! This PR is an artifact of our previous, now-abandoned strategy for how persistence was going to plug into coord+dataflow and is no longer necessary.

**Comment by frankmcsherry** [comment]
Should we close it then?

**Comment by danhhz** [comment]
I have no opinion either way. I have a very vague recollection of Nikhil saying that this was how he wished it had been done in the first place (but I might be confusing it for something else), so I'll let him handle that question

**Comment by benesch** [comment]
Oops, sorry for missing the question the first time around. 

> It seems maybe weird to oblige the `coord <-> dataflow` API to reflect the intended persistence quantum. Could be polite, for sure, but what is the constraint here / why are we changing this API w/o review from coord / dataflow folks?

Right, yeah, that's why this PR got abandoned. It predated the platform/architecture discussions by a bit.

**Comment by benesch** [comment]
> I have no opinion either way. I have a very vague recollection of Nikhil saying that this was how he wished it had been done in the first place (but I might be confusing it for something else), so I'll let him handle that question

Yeah, I do think it's a bit funny that you can't bundle inserts to multiple tables in one command, but no strong feels if we're not saying anything about the atomicity of a `SequencedCommand::Insert` in the new design.

**Pull request closed without merging**